src/java/org/apache/cassandra/repair/autorepair/RepairRangeSplitter.java

```diff
@@ -0,0 +1,408 @@
+
+    @VisibleForTesting
+    static RepairAssignment merge(List<RepairAssignment> assignments)
+    {
+        if (assignments.isEmpty())
+            throw new IllegalStateException("Cannot merge empty assignments");
+
+        Set<String> mergedTableNames = new HashSet<>();
+        Range<Token> referenceTokenRange = assignments.get(0).getTokenRange();
+        String referenceKeyspaceName = assignments.get(0).getKeyspaceName();
+
+        for (RepairAssignment assignment : assignments)
+        {
+            // These checks _should_ be unnecessary but are here to ensure that the assignments are consistent
+            if (!assignment.getTokenRange().equals(referenceTokenRange))
+                throw new IllegalStateException("All assignments must have the same token range");
+            if (!assignment.getKeyspaceName().equals(referenceKeyspaceName))
+                throw new IllegalStateException("All assignments must have the same keyspace name");
+
+            mergedTableNames.addAll(assignment.getTableNames());
+        }
+
+        return new RepairAssignment(referenceTokenRange, referenceKeyspaceName, new ArrayList<>(mergedTableNames));
+    }
+
+    public List<RepairAssignment> getRepairAssignmentsForTable(AutoRepairConfig.RepairType repairType, String keyspaceName, String tableName, Collection<Range<Token>> tokenRanges)
+    {
+        List<RepairAssignment> repairAssignments = new ArrayList<>();
+
+        long targetBytesSoFar = 0;
+
+        List<SizeEstimate> sizeEstimates = getRangeSizeEstimate(repairType, keyspaceName, tableName, tokenRanges);
+        // since its possible for us to hit maxBytesPerScheduleBytes before seeing all ranges, shuffle so there is chance
+        // at least of hitting all the ranges _eventually_ for the worst case scenarios
+        Collections.shuffle(sizeEstimates);
+        for (SizeEstimate estimate : sizeEstimates)
+        {
+            if (estimate.sizeInRange == 0)
+            {
+                ColumnFamilyStore cfs = ColumnFamilyStore.getIfExists(keyspaceName, tableName);
+                long memtableSize = cfs.getTracker().getView().getCurrentMemtable().getLiveDataSize();
+                if (memtableSize > 0L)
+                {
+                    logger.debug("Included {}.{} range {}, had no unrepaired SSTables, but memtableSize={}, adding single repair assignment", keyspaceName, tableName, estimate.tokenRange, memtableSize);
+                    RepairAssignment assignment = new RepairAssignment(estimate.tokenRange, keyspaceName, Collections.singletonList(tableName));
+                    repairAssignments.add(assignment);
+                }
+                else
+                {
+                    logger.debug("Skipping {}.{} for range {} because it had no unrepaired SSTables and no memtable data", keyspaceName, tableName, estimate.tokenRange);
+                }
+            }
+            else if (targetBytesSoFar + estimate.sizeInRange < maxBytesPerScheduleBytes)
```

Comment:
Thought about this in the prototype and realized it's probably better to use `totalSize`  for incremental as it more accurately represents then work (since SSTable would need to be anticompacted anyways in incremental case), but maybe for Full repair, sizeInRange makes senseas it currently is.